server/src/main/java/org/elasticsearch/common/util/CancellableSingleObjectCache.java

```diff
@@ -0,0 +1,277 @@
+                singleUseCachedItem.decRef();
+                return;
+            }
+
+            if (newCachedItem == null) {
+                newCachedItem = new CachedItem(key);
+            }
+
+            if (currentCachedItemRef.compareAndSet(currentCachedItem, newCachedItem)) {
+                if (currentCachedItem != null) {
+                    currentCachedItem.decRef();
+                }
+                startRefresh(key, newCachedItem);
+                final boolean listenerAdded = newCachedItem.addListener(listener, isCancelled);
+                assert listenerAdded;
+                newCachedItem.decRef();
+                return;
+            }
+            // else the CAS failed because we lost a race to a concurrent retrieval; try again from the top since we expect the race winner
+            // to be fresh enough for us and therefore we can just wait for its result.
+        } while (true);
+    }
+
+    private void startRefresh(Key key, CachedItem cachedItem) {
+        try {
+            refresh(key, cachedItem::ensureNotCancelled, cachedItem.getFuture());
+        } catch (Exception e) {
+            cachedItem.getFuture().onFailure(e);
+        }
+    }
+
+    /**
+     * An item in the cache, representing a single invocation of {@link #refresh}.
+     * <p>
+     * This item is ref-counted so that it can be cancelled if it becomes irrelevant. References are held by:
+     * <ul>
+     *     <li>Every listener that is waiting for the result, released on cancellation. There's no need to release on completion because
+     *     there's nothing to cancel once the refresh has completed.</li>
+     *     <li>The cache itself, released once this item is no longer the current one in the cache, either because it failed or because a
+     *     fresher computation was started.</li>
+     *     <li>The process that adds the first listener, released once the first listener is added.</li>
+     * </ul>
+     */
+    private final class CachedItem extends AbstractRefCounted {
+
+        private final Key key;
+        private final PlainListenableActionFuture<Value> future = PlainListenableActionFuture.newListenableFuture();
+        private final CancellationChecks cancellationChecks = new CancellationChecks();
+
+        CachedItem(Key key) {
+            super("cached item");
+            this.key = key;
+            incRef(); // start with a refcount of 2 so we're not closed while adding the first listener
+            this.future.addListener(new ActionListener<>() {
+                @Override
+                public void onResponse(Value value) {
+                    cancellationChecks.clear();
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    cancellationChecks.clear();
+                    // Do not cache this failure
+                    if (currentCachedItemRef.compareAndSet(CachedItem.this, null)) {
+                        decRef();
```

Comment:
I *think* we have this here to make `addListener` return false if it can, but I am not really sure it is the only reason?

Reply:
Mainly it's there because that's what the comment that describes the ref-counting discipline says :) Yes this only really has the effect of allowing any concurrently-added listeners to retry rather than inheriting the failure. I added a comment to that effect in 4512be8b9029eeba40a4d78b0ed342627d3eafcd.